pageserver/src/walredo.rs

```diff
@@ -187,10 +188,39 @@ impl PostgresRedoManager {
         PostgresRedoManager {
             tenant_id,
             conf,
+            last_redo_at: std::sync::Mutex::default(),
             redo_process: RwLock::new(None),
         }
     }
 
+    /// This type doesn't have its own background task to check for idleness: we
+    /// rely on our owner calling this function periodically in its own housekeeping
+    /// loops.
+    pub(crate) fn maybe_quiesce(&self, idle_timeout: Duration) {
+        use std::sync::TryLockError;
+
+        {
+            let g = match self.last_redo_at.try_lock() {
+                Ok(g) => g,
+                Err(TryLockError::Poisoned(e)) => e.into_inner(),
+                Err(TryLockError::WouldBlock) => return,
+            };
+
+            match *g {
+                None => return,
+                Some(last_redo_at) if last_redo_at.elapsed() < idle_timeout => {
+                    return;
+                }
+                Some(_too_long_ago) => {
+                    // fallthrough
+                }
+            }
+        }
+
+        let mut guard = self.redo_process.write().unwrap();
+        *guard = None;
+    }
```

Comment:
```suggestion
    pub(crate) fn maybe_quiesce(&self, idle_timeout: Duration) {
    use std::sync::TryLockError;

    if let Ok(g) = self.last_redo_at.try_lock() {
        if let Some(last_redo_at) = *g {
            if last_redo_at.elapsed() >= idle_timeout {
                let mut guard = self.redo_process.write().unwrap();
                *guard = None;
            }
        }
    }
}
```

Reply:
it's easier for me to read this way, as the function doesn't do  anything with TryLockError::Poisoned(e) and TryLockError::WouldBlock

Reply:
> doesn't do anything with TryLockError::Poisoned(e)

Actually the poisoned value was accepted but I can see that your version is much simpler.

I'll hack in the "drop mutex before aquiring writelock" to your solution.

Reply:
[c8dc861](https://github.com/neondatabase/neon/pull/5877/commits/c8dc861dc027ba5e26a3d84815e295d4a4aa21ba)